feat: add WebAuthn dialog support on macOS

EricRahm opened the pull request:
#### Description of Change
This PR adds support for WebAuthn dialog on macOS.

Closes #27355

Please note this is an revived and updated version of #28349.

#### Checklist
- [ ] PR description included and stakeholders cc'd
- [ ] `npm test` passes
- [ ] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes
Notes: Added support for WebAuthn on macOS

zcbenz commented on shell/browser/electron_browser_client.cc:
```diff
@@ -112,6 +112,9 @@
 #include "v8/include/v8.h"
 
+#include "chrome/browser/webauthn/authenticator_request_scheduler.h"
+#include "chrome/browser/webauthn/chrome_authenticator_request_delegate.h"
```
Can you move the 2 lines to the list above? Headers generally are only listed separated when they are conditional.

EricRahm replied:
Move them up.

zcbenz commented on shell/browser/electron_browser_context.h:
```diff
@@ -81,6 +81,9 @@ class ElectronBrowserContext : public content::BrowserContext {
       bool in_memory,
       base::DictionaryValue options = base::DictionaryValue());
 
+  static ElectronBrowserContext* FromBrowserContext(
```
We don't usually add helpers when it is just a simple `static_cast`.

EricRahm replied:
I've gone with the `static_cast` and removed the helper.

zcbenz reviewed (approved):
API LGTM

EricRahm commented:
@zcbenz Thanks for the quick review! I'll update the PR later today with your suggestions.

EricRahm commented on patches/chromium/fix_add_support_for_webauthn_on_macos.patch:
```diff
@@ -0,0 +1,322 @@
+   }
+ #endif
+ 
+-  // Chrome disables platform authenticators is Guest sessions. They may be
```
Makes sense, I've updated the patch to `#if 0` the deleted lines instead.

EricRahm commented:
@zcbenz I've updated the PR with the changes you requested. I rebased against the lastest main as well to resolve conflicts that would have blocked landing.

EricRahm commented:
FWIW this backports cleanly with one minor tweak to the electron/16-x-y branch.

zcbenz commented:
/cc @electron/wg-upgrades @electron/wg-api

poiru commented:
Pinging @electron/wg-upgrades @electron/wg-api again after the December break.

nornagon commented on patches/chromium/fix_add_support_for_webauthn_on_macos.patch:
```diff
@@ -0,0 +1,315 @@
+       const int omnibox_primary_delta =
+           GetFontSizeDeltaBoundedByAvailableHeight(
+-              LocationBarView::GetAvailableTextHeight(),
++              // LocationBarView::GetAvailableTextHeight(),
```
I feel like the positioning of the dialog should probably be customizable by apps? Can we get some screenshots of how this looks?

EricRahm replied:
I believe that's the point of `marginTop`?

![image](https://user-images.githubusercontent.com/426919/150265606-98d765e9-87f4-49fd-9f80-ca328837aaf5.png)

![image](https://user-images.githubusercontent.com/426919/150265652-b25801af-f633-4e2e-a098-6aa820999128.png)

EricRahm replied:
@nornagon Would you be okay with focusing on the core functionality in this PR and iterating on further customization in a follow up?

EricRahm replied:
This is with:
```TypeScript
mainWindow.setModalDialogOptions({ marginTop: 100 });
```
![image](https://user-images.githubusercontent.com/426919/150266686-329bf7b9-d3a6-4f15-840a-4064db5f6a01.png)

deepak1556 commented on chromium_src/BUILD.gn:
```diff
@@ -4,14 +4,32 @@
+copy("shaded_resources_copy") {
+  deps = [ "//electron:resources" ]
+  sources = [ "$root_gen_dir/electron/grit/electron_resources.h" ]
+  outputs = [ "$target_gen_dir/chrome/grit/generated_resources.h" ]
```
`patches/chromium/fix_add_support_for_webauthn_on_macos.patch ` already uses `electron/grit/electron_resources.h` where needed, why is this step necessary to copy contents to `chrome/grit/generated_resources.h` ?

EricRahm replied:
Good question, I'll see if I can remove it.

EricRahm replied:
@deepak1556 Removing that code worked fine, I've pushed a fix for that.

nornagon commented on patches/chromium/fix_add_support_for_webauthn_on_macos.patch:
```diff
@@ -0,0 +1,315 @@
+-  const Profile* profile = Profile::FromBrowserContext(browser_context);
+-  const PrefService* prefs = profile->GetPrefs();
++  const auto* profile = static_cast<ElectronBrowserContext*>(browser_context);
++  const PrefService* prefs = profile->prefs();
```
🤔  minor: perhaps we should add a `GetPrefs()` method to `ElectronBrowserContext` to slightly reduce the patch surface area here?

EricRahm replied:
Seems reasonable

EricRahm replied:
I want ahead and did this in a03f822, compiles fine and reduces the chromium patch a bit.

nornagon commented on patches/chromium/fix_add_support_for_webauthn_on_macos.patch:
```diff
@@ -0,0 +1,315 @@
+   return base::UTF16ToUTF8(nonbreaking_name);
+ }
+ 
++#if 0
+ // PairingFromSyncedDevice extracts the caBLEv2 information from Sync's
+ // DeviceInfo (if any) into a caBLEv2 pairing. It may return nullptr.
```
TIL https://blog.millerti.me/2021/06/18/previewing-chromes-cable-v2-support-for-webauthn/

obvs "cloud-assisted" isn't something we can support in Electron, but this is neat!

nornagon commented on docs/api/browser-window.md:
```diff
@@ -1810,6 +1810,13 @@ with `addBrowserView` or `setBrowserView`.
 **Note:** The BrowserView API is currently experimental and may change or be
 removed in future Electron releases.
 
+#### `win.setModalDialogOptions(options)`
```
I'd like to bikeshed this name a bit. `modal` is a concept that has quite a lot of meanings already, in particular, `BrowserWindow` can be constructed with a `modal: true` parameter, which this method is completely unrelated to.

How about `win.setWebAuthnDialogOptions`? Do we expect to implement other overlay dialogs like this one in the future?

EricRahm replied:
This sets an attribute on a generic `ModalDialogHost` owned by `NativeWindow`. WebAuthn is currently the only usage of that, but it's a generic component and could certainly be used by other features. I don't feel strongly about any of this and am happy to name it whatever y'all feels is best.

nornagon commented:
Also, do you think it would be possible to write some automated tests for this functionality? It's quite a large feature and I'd like it to at least have _some_ test coverage.

EricRahm commented:
> Also, do you think it would be possible to write some automated tests for this functionality? It's quite a large feature and I'd like it to at least have _some_ test coverage.

I'm new to the project so I'm not totally sure how UI interaction tests that require a specific piece of hardware to be inserted and physically tapped would work in automation, can you elaborate on what you'd propose here?

poiru commented:
It's possible you could use the virtual authenticator environment by exposing the relevant APIs to the test JS code. See:
* https://developer.chrome.com/docs/devtools/webauthn/
* `AuthenticatorEnvironmentImpl::EnableVirtualAuthenticatorFor` and `VirtualAuthenticatorManagerImpl::AddAuthenticatorAndReturnNonOwningPointer` in Chromium

EricRahm commented:
> It's possible you could use the virtual authenticator environment by exposing the relevant APIs to the test JS code. See:
> 
> * https://developer.chrome.com/docs/devtools/webauthn/
> * `AuthenticatorEnvironmentImpl::EnableVirtualAuthenticatorFor` and `VirtualAuthenticatorManagerImpl::AddAuthenticatorAndReturnNonOwningPointer` in Chromium

The scope of those are limited to devtools and content shell (for WPT), I'm not sure we can (or event want to) expose them. They completely bypass any UI which is a lot of what this patch is about. @nornagon Do you have any thoughts here?

nornagon commented:
I don't suppose we can use the non-USB version?
![Screen Shot 2022-01-26 at 4 00 08 PM](https://user-images.githubusercontent.com/172800/151266997-e0367177-baf5-496b-a653-b7a75b5c69fc.png)

EricRahm commented:
In Chrome, that's going to spinup a platform specific *system* dialog. For example on macos that's going ask for a system password or for devices that have it your fingerprint. We haven't wired in support for that in Electron, so you'll just see a sad dialog:

<img width="743" alt="image" src="https://user-images.githubusercontent.com/426919/151268842-ef0c425b-e099-4179-972d-2946e8edc516.png">

dsanders11 commented on shell/browser/ui/views/electron_constrained_window_views_client.cc:
```diff
@@ -0,0 +1,46 @@
+    return platform_util::GetViewForWindow(parent);
+  }
+
+  DISALLOW_COPY_AND_ASSIGN(ElectronConstrainedWindowViewsClient);
```
The Electron codebase no longer uses `DISALLOW_COPY_AND_ASSIGN` as of #31633. Could you update this PR to change these usages out?

EricRahm replied:
Yep I'll swap em out.

EricRahm replied:
Updated in 315f996

dsanders11 replied:
Could you also remove the includes of "base/macros.h"? Those were only needed for `DISALLOW_COPY_AND_ASSIGN`.

EricRahm replied:
Yep!

dsanders11 commented on shell/browser/native_window.cc:
```diff
@@ -11,19 +11,28 @@
+#include "ui/gfx/geometry/rect.h"
+#include "ui/gfx/geometry/size.h"
 #include "ui/views/widget/widget.h"
+#include "ui/views/window/client_view.h"
```
I don't think this include is used? Unless I'm missing the usage.

dsanders11 commented on shell/browser/ui/electron_web_modal_dialog_manager_delegate.h:
```diff
@@ -0,0 +1,29 @@
+#ifndef ELECTRON_SHELL_BROWSER_UI_ELECTRON_WEB_MODAL_DIALOG_MANAGER_DELEGATE_H_
+#define ELECTRON_SHELL_BROWSER_UI_ELECTRON_WEB_MODAL_DIALOG_MANAGER_DELEGATE_H_
+
+#include "base/compiler_specific.h"
```
Pretty sure this include isn't used.

dsanders11 commented on shell/browser/native_window.cc:
```diff
@@ -11,19 +11,28 @@
 #include "ui/display/win/screen_win.h"
 #endif
 
+using views::View;
```
This doesn't look to be used in this file. I see it's from upstream, but since it isn't used here, let's drop it.

dsanders11 commented on shell/browser/ui/views/electron_constrained_window_views_client.cc:
```diff
@@ -0,0 +1,50 @@
+
+std::unique_ptr<constrained_window::ConstrainedWindowViewsClient>
+CreateElectronConstrainedWindowViewsClient() {
+  return std::make_unique<ElectronConstrainedWindowViewsClient>();
```
Add `#include <memory>` to this file for this usage.

sneko commented:
Hi,

Thanks for you work on this PR! Can someone tell me where we stand to have this merged?

sneko commented:
Hi @MarshallOfSound , just saw you got assigned to this one.

Do you know what are the remaining steps to have this merged?

Thank you :)

saenzramiro commented:
I'm very excited about this PR. 😁

saenzramiro commented:
Any news on this?

sneko commented:
I also definitely want to understand what makes this PR stuck? Just to know if we can expect to have it in our Electron apps soon :)

Thank you,

MarshallOfSound commented:
@EricRahm Can you please rebase this PR on latest `main`, there are changes there that should make this PR easier to take (namely the strings are no longer copied into our source).  I'll give this a closer look once it has been rebased and CI is looking a bit happier.

EricRahm commented:
> @EricRahm Can you please rebase this PR on latest `main`, there are changes there that should make this PR easier to take (namely the strings are no longer copied into our source). I'll give this a closer look once it has been rebased and CI is looking a bit happier.

@MarshallOfSound Apologies, enough time has passed that I'm now working for a new company and don't have time to dedicate to this (nor do I have a working setup for hacking on Electron). *If someone wants to take this over and pursue a rebase they are more than welcome to.* It's possible I'll have time to take another look in the coming months (I'll also need to confirm I can contribute at all).

juancampa commented:
I'll be working on landing this PR in the next few days. I'm taking over @EricRahm's awesome work but I'll need a bit of time to catch up. I'll get this rebased as soon as possible.

AramAnd commented:
+1

sneko commented:
@zcbenz thank you for approving this PR. Do you know why the UI shows you are still pending for review? Maybe the UI shows it as an history "pending, then approved" but it's not clear.

<img width="351" alt="image" src="https://user-images.githubusercontent.com/8195958/182102518-b0e0df26-1bf8-4251-a18c-77c2d256951e.png">


<img width="573" alt="image" src="https://user-images.githubusercontent.com/8195958/182102265-ac011fb3-4b8b-4bc8-b28f-6eac27842ac0.png">

Thank you,

sneko commented:
Hi, do you know if that's something on the Electron roadmap?

I'm a bit surprised there is so little consideration from users (thumbs up...) whereas Electron apps are everywhere and that WebAuthn is really important for security.

Maybe the explanation is in most cases the auth flows are done in the desktop browser to manage WebAuthn and injected into the Electron app?

saenzramiro commented:
@MarshallOfSound @zcbenz Please, consider this PR as important, is related with security.

juancampa commented:
I want to clarify that this PR needs a new owner. The company that was sponsoring the development of this feature doesn't exist anymore, sorry about that.

But hey! If you're interested, feel free to chime in and take it to the finish line, it's not that far from it IMO.

EricRahm commented:
I think it's time to put this PR to bed, if anyone wants to resurrect it I would recommend fighting that urge.

SagaciousLittle commented:
Does anyone know why this is an issue specific to electron mac? chrome mac doesn't seem to have this problem?
